Add expiration to master keys & encryption subkey

trishankatdatadog opened the pull request:
* Add expiration to master keys and expiration subkeys
* Check for expiration of master key when returning encryption/signing subkeys
* Find newest instead of first unexpired signing subkey

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog commented:
I suppose there could be a slight difference in the creation and expiration time of the master key and encryption subkey now, but it should not matter in practice.

twiss commented:
Hey @trishankatdatadog, thanks for the PR!

Typically, key expiry is set on the primary key only. Of course subkeys _can_ expire separately, but most typically, the subkeys expire when the primary key expires. (This makes it easier to extend the lifetime of the entire key, since you just have to re-sign one signature.)
Unfortunately, this is not how ProtonMail/crypto currently behaves; subkeys' lifetime currently extends beyond the lifetime of the primary key, but this is a bug. (See https://github.com/ProtonMail/gopenpgp/issues/80, which has this as its root cause.) So probably [EncryptionKey](https://github.com/ProtonMail/crypto/blob/aef62243dce8452451d6dd6c009b22c58a6a50fa/openpgp/keys.go#L91) should check whether the primary key is expired _before_ looking at the subkeys. (And similarly for [SigningKey](https://github.com/ProtonMail/crypto/blob/aef62243dce8452451d6dd6c009b22c58a6a50fa/openpgp/keys.go#L129).) And then, personally I would leave the subkey generation code as it was, to avoid setting an expiration there and also to avoid the (admittedly minor) potential issue of different creation dates.

trishankatdatadog commented:
@twiss, a few comments:

1. So I fixed the code to return no signing/encryption keys when the primary UID has expired. However, I also added a test to show that this fails with a direct signature on the primary key itself has expired. AFAICT, we do not handle such signature packets at all. How should we handle this?

2. I brought back the old code to directly add the encryption subkey. I still find it a little inconsistent to not add expiry on this subkey, but checking the expiration of the primary UID does mitigate this.

3. I just noticed that we get the _newest_ unexpired encryption subkey, but the _first_ unexpired signing subkey. What is the rationale for this? The latter can lead to unintuitive behavior from an end-user's perspective.

trishankatdatadog commented:
_ping_ 🙂

twiss commented:
Hey @trishankatdatadog, sorry for the delay, and thanks for the ping :)

1. IMO, I wouldn't worry about this for now. This is an uncommon way of expiring a key. So I would just leave out this test for now.
2. Yeah, I understand. This is standard for OpenPGP keys, though
3. Yeah, good point. It should return the newest in both cases indeed. Feel free to fix it here or in a separate PR, if you feel like :)

twiss commented:
@trishankatdatadog Just for clarity, and as a gentle ping, I'm happy to merge this without the failing test :blush:

trishankatdatadog commented:
> @trishankatdatadog Just for clarity, and as a gentle ping, I'm happy to merge this without the failing test 😊

Oh, hey, sorry, lemme get around to this, was busy with other projects, thanks!

twiss commented:
Np, thanks! :pray:

trishankatdatadog commented:
@twiss Reverted the bad test, let me get around to returning the first unexpired signing subkey ASAP in this PR?

twiss commented:
@trishankatdatadog Yeah, of course, thanks! :pray:

trishankatdatadog commented:
Working on testing finding the newest instead of the first unexpired signing subkey...

trishankatdatadog commented:
@twiss Done, please take a careful look. In particular, I found that I had to add [this line](https://github.com/ProtonMail/crypto/pull/64/files#diff-f9f6fc7d09d46d44a33548b3d8c193bb88251597f0fed6d4b6d25670715ca1cfR820) to prevent keys from being considered immediately expired, but the [corresponding line](https://github.com/ProtonMail/crypto/pull/64/files#diff-65b30b147e6c8432806b9ff7fa5f3368af59ccbd942e99582137fa0cd46978f6R524) seems to have no effect on whether signatures are considered immediately expired (they are not).

twiss commented on openpgp/keys.go:
```diff
@@ -106,53 +112,64 @@ func (e *Entity) EncryptionKey(now time.Time) (Key, bool) {
 		subkey := e.Subkeys[candidateSubkey]
-		return Key{e, subkey.PublicKey, subkey.PrivateKey, subkey.Sig}, true
+		encryptionKey, found = Key{e, subkey.PublicKey, subkey.PrivateKey, subkey.Sig}, true
+		return
```
To me, the old pattern of returning directly, rather than assigning and then returning, is a little bit more clear. Could you revert these changes? (Also in `SigningKey`)

twiss commented on openpgp/packet/signature.go:
```diff
@@ -935,7 +935,7 @@ func (sig *Signature) AddMetadataToHashSuffix() {
 	l := uint64(
 		uint64(n[0])<<56 | uint64(n[1])<<48 | uint64(n[2])<<40 | uint64(n[3])<<32 |
-		uint64(n[4])<<24 | uint64(n[5])<<16 | uint64(n[6])<<8  | uint64(n[7]))
+			uint64(n[4])<<24 | uint64(n[5])<<16 | uint64(n[6])<<8 | uint64(n[7]))
```
I assume this was autoformatted, but the old spacing was a little bit clearer to me

twiss reviewed (commented):
@trishankatdatadog Thanks! Yeah, makes sense. That's probably because `SigExpired` is not called as often, but yes this change seems correct. I've left two very minor nitpicks, but other than that looks good to me :+1:

trishankatdatadog commented:
> @trishankatdatadog Thanks! Yeah, makes sense. That's probably because `SigExpired` is not called as often, but yes this change seems correct. I've left two very minor nitpicks, but other than that looks good to me 👍

Done, please take a final look 👀

twiss reviewed (approved):
Looks good, thanks a lot! :+1:
